Add exporter to a local MongoDB

amanchoudhri opened the pull request:
Add functions to export a given graph to a local MongoDB and then query some basic information back from that DB.

amanchoudhri committed:
Add the schema validation functions to __init__

amanchoudhri committed:
Require pymongo with pybel installation

amanchoudhri committed:
Create exporter and basic query functions

amanchoudhri committed:
Fix imports

amanchoudhri committed:
Reorder imports

amanchoudhri committed:
Break up get_edges into two smaller functions

Before, get_edges() could be called either with:
a node as an input or criteria to search for matching nodes as the input
Now, those two input sets are broken up into smaller functions

amanchoudhri committed:
Pass flake8

cthoyt commented:
Needs unit tests!

amanchoudhri committed:
Restructure exporter from a folder to a file

amanchoudhri committed:
Start writing tests

amanchoudhri committed:
Add test for edge presence in collection

amanchoudhri committed:
Add test for find_nodes() query function

amanchoudhri committed:
Start a mongodb server during Travis testing

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,152 @@
+    :return: the collection that now stores the graph
+    """
+    # Access (or create) the specified database and collection
+    db = client[db_name]
```
make the client a keyword argument that's optional:
```python
def to_mongodb(graph: pybel.BELGraph, db_name: str, collection_name: str, client: Optional[MongoClient] = None) -> Collection:
    if client is None:
        client = MongoClient()
    ...
```

This way, the user has the ability to specify stuff if they want, and we don't have to worry about the client as a global variable to the module (will help with testing)

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,152 @@
+    collection: Collection,
+    name: str = None,
+    identifier: str = None,
+    variants: List[pybel.dsl.EntityVariant] = None,
```
should just be `pybel.dsl.Variant`

cthoyt commented on tests/test_io/test_mongodb.py:
```diff
@@ -0,0 +1,120 @@
+class TestMongoDB(unittest.TestCase):
+    """Tests for the MongoDB exporting and querying."""
+    def setUp(self):
+        """Create and export a test graph."""
```
create an instance of the mongo client here, and if it fails, use `self.skip()` - this will make it pass on appveyor, where mongodb is not available

cthoyt commented on tests/test_io/test_mongodb.py:
```diff
@@ -0,0 +1,120 @@
+        assert_fn = self.assertEqual
+        if not assert_in:
+            assert_fn = self.assertNotEqual
+        if type(edge) is not dict:
```
use `isinstance`

amanchoudhri committed:
Relocate mongo server connection to to_mongodb()

amanchoudhri committed:
Add mongo connection to setUp, skip tests on error

amanchoudhri committed:
Add function to remove MongoDB key entries

Before, there was a lot of repetition in deleting the '_id' and
'type' keys from mongo items. This commit makes that more DRY by
pulling that functionality out into _rm_mongo_keys().

amanchoudhri committed:
Add tests for getting edges from node/criteria

amanchoudhri committed:
Remove extraneous 'self' param from _rm_mongo_keys

amanchoudhri committed:
Finalize tests for retrieving edges

amanchoudhri committed:
Merge branch 'master' into add-mongo-exporter

amanchoudhri committed:
Remove unnecessary _handle_member function

amanchoudhri committed:
Debug and rework edge tests for SBEL data format

amanchoudhri committed:
Merge branch 'add-mongo-exporter' of https://github.com/pybel/pybel into add-mongo-exporter

amanchoudhri committed:
Clean up imports

cthoyt committed:
Make SBEL more functional

amanchoudhri committed:
Merge 'add-mongo-exporter' into branch

amanchoudhri committed:
Modify test to include new to_sbel functionality

amanchoudhri committed:
Add option to search for nodes based on function

amanchoudhri committed:
Pass flake8

amanchoudhri committed:
Merge branch 'master' into add-mongo-exporter

cthoyt commented on src/pybel/__init__.py:
```diff
@@ -45,6 +45,7 @@
 from .io.umbrella_nodelink import to_umbrella_nodelink, to_umbrella_nodelink_file, to_umbrella_nodelink_gz
 from .manager import Manager, from_database, to_database
 from .parser.parse_bel import parse
+from .schema import is_valid_edge, is_valid_node
```
I'd rather this weren't exposed at the top level. Anyone who wants to get into the nitty-gritty of the schema should be comfortable importing it with `from pybel.schema import is_valid_edge, is_valid_node`

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,173 @@
+from pymongo import MongoClient
+from pymongo.collection import Collection
+
+from pybel import BELGraph, to_sbel
```
relative imports

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,173 @@
+logger = logging.getLogger(__name__)
+
+
+def _entity_to_dict(entity: Entity) -> Mapping[str, Any]:
```
why is this necessary? isn't the Entity already a dict?

aman527 replied:
Yep, that makes sense. I just need it in testing to make sure the node is equal to the Mongo entry, so the two need to be of the exact same type. I'll relocate it to the test file for clarity.

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,173 @@
+def _entity_to_dict(entity: Entity) -> Mapping[str, Any]:
+    """Input a pybel Entity and return a dict representing it."""
+    new_node = dict(entity)
+    if new_node['function'] in ['Complex', 'Composite']:
```
you should use the constants representing this strings in `pybel.constants`:
`pybel.constants.FUNCTION`, `pybel.constants.COMPLEX`, `pybel.constants.COMPOSITE`, `pybel.constants.MEMBERS`, etc.

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,173 @@
+
+def to_mongodb(
+    graph: BELGraph,
+    db_name: str, collection_name: str,
```
would you please choose a default `db_name` and `collection_name` in case the user just wants reasonable defaults?

Save these as constants at the top of the file, then make these two arguments`Optional[str] = None`. Inside the function, you can do the same none check as with the client

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,173 @@
+    # Access (or create) the specified database and collection
+    db = client[db_name]
+    collection = db[collection_name]
+    # If a collection with the same name already exists, drop its contents
```
why? what if someone uploads one graph, then later wants to upload another graph? or should the collection correspond to the graph?

aman527 replied:
My thinking was that each collection would correspond to one graph.

cthoyt replied:
is it possible to query over all collections simultaneously?

aman527 replied:
I'm not sure, but I don't think so

cthoyt replied:
Did you get a chance to look into this?

aman527 replied:
I could write a function to query over all collections, but it would just be a loop, not simultaneous

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,173 @@
+        n = deepcopy(node)
+        n = _entity_to_dict(n)
+        n['type'] = 'node'
+        collection.insert_one(n)
```
is there a bulk insertion option for the collection? It will likely be faster

cthoyt commented on tests/test_io/test_mongodb.py:
```diff
@@ -0,0 +1,216 @@
+NodeInfo = namedtuple('NodeInfo', 'name identifier function variants')
+
+
+def id_info(node: Mapping[str, Any]) -> NodeInfo:
```
I don't see why you need the NodeInfo object, can't you access all of the name, identifier, function, and variants directly from the node object anyway?

aman527 replied:
Yep I can, but having the NodeInfo object reduces the repetition of .get() calls on node objects and makes it easier to add criteria (like the function param I added yesterday). Does that make it worth keeping, or does it just add clutter/complexity?

cthoyt commented on tests/test_io/test_mongodb.py:
```diff
@@ -0,0 +1,216 @@
+        matches_from_criteria = []
+        # Since get_edges_from_criteria can return edges for multiple matching nodes,
+        # loop through every tuple (node, [edges...]) in pairs_from_criteria
+        for (node_match, edges_match) in pairs_from_criteria:
```
no parentheses necessary when unpacking in a for loop

amanchoudhri committed:
Remove schema imports from top level

amanchoudhri committed:
Make imports relative

amanchoudhri committed:
Relocate _entity_to_dict() to the tests file

amanchoudhri committed:
Replace string literals with pybel constants

amanchoudhri committed:
Add default db/collection and bulk insert entries

amanchoudhri committed:
Merge branch 'add-mongo-exporter' of https://github.com/pybel/pybel into add-mongo-exporter

amanchoudhri committed:
Pass flake8

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,203 @@
+     Note: nodes that contain the variants in addition to specified variants will be matched.
+    :return: A list containing all the nodes that match the given criteria
+    """
+    if not (name or identifier):
```
an identifier is pretty useless without a namespace, isn't it?

aman527 replied:
Should I remove the identifier option or add in a namespace param?

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,203 @@
+    :param name: The name of the desired node
+    :param identifier: The identifier of the desired node
+    :param function: The type of the desired node ("protein", "complex", etc)
+    :param variants: A list of variants that the desired node should contain.
```
docstrings don't match actual arguments

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,203 @@
+        function=node_function,
+        variants=node_variants
+    )
+    edges = []
```
would a dictionary data structure work better? Nodes are hashable, so dict.items() would get you effectively the same thing as this too

cthoyt committed:
Add flake8 check to tests

cthoyt commented on src/pybel/io/mongodb.py:
```diff
@@ -0,0 +1,204 @@
+        if not value:
+            del filter_[key]
+
+    return list(collection.find(filter_))
```
can you turn these into PyBEL entities using `pybel.tokens.parse_result_to_dsl`? If the nodes are in the right structure, you can do:
```python
return [
    pybel.tokens.parse_result_to_dsl(node)
    for node in collection.find(filter_)
]
```

cthoyt committed:
Merge branch 'master' into add-mongo-exporter

cthoyt committed:
Fix names

cthoyt committed:
Return nodes as pybel classes

cthoyt committed:
Update test_mongodb.py

amanchoudhri committed:
Parse found nodes to dsl

amanchoudhri committed:
Merge branch 'add-mongo-exporter' of https://github.com/pybel/pybel into add-mongo-exporter

amanchoudhri committed:
Parse results of found_node() to DSL entities

In the previous "Parse found nodes to dsl" commit tests were failing.
This commit adds a new function to ensure that querying for edges still
works as expected, and all mongodb tests now pass.

amanchoudhri committed:
Pass flake8

amanchoudhri committed:
Pass flake8

amanchoudhri committed:
Improve documentation and add code examples

amanchoudhri committed:
Fix loop where a dict changes size mid-iteration
